atomics/include/desul/atomics/Compare_Exchange_OpenACC.hpp

```diff
@@ -0,0 +1,145 @@
+/*
+Copyright (c) 2019, Lawrence Livermore National Security, LLC
+and DESUL project contributors. See the COPYRIGHT file for details.
+Source: https://github.com/desul/desul
+
+SPDX-License-Identifier: (BSD-3-Clause)
+*/
+
+#ifndef DESUL_ATOMICS_COMPARE_EXCHANGE_OPENACC_HPP_
+#define DESUL_ATOMICS_COMPARE_EXCHANGE_OPENACC_HPP_
+
+#include <openacc.h>
+
+#include <desul/atomics/Common.hpp>
+#include <desul/atomics/Thread_Fence_OpenACC.hpp>
+#include <type_traits>
+
+namespace desul {
+namespace Impl {
+
+#ifdef __NVCOMPILER
+
+#pragma acc routine seq
+template <class T, class MemoryOrder, class MemoryScope>
+T device_atomic_exchange(T* dest, T value, MemoryOrder, MemoryScope /*scope*/) {
+  if constexpr (std::is_arithmetic_v<T> && ((sizeof(T) == 4) || (sizeof(T) == 8))) {
+    T return_val;
+#pragma acc atomic capture
+    {
+      return_val = *dest;
+      *dest = value;
+    }
+    return return_val;
+  } else {
+    // FIXME_OPENACC
+    printf(
+        "DESUL error in device_atomic_exchange(): Not supported atomic operation in "
+        "the OpenACC backend\n");
+    //  Acquire a lock for the address
+    // while (!lock_address_openacc((void*)dest, scope)) {
+    // }
+    // device_atomic_thread_fence(MemoryOrderAcquire(), scope);
+    T return_val = *dest;
+    *dest = value;
+    // device_atomic_thread_fence(MemoryOrderRelease(), scope);
+    // unlock_address_openacc((void*)dest, scope);
+    return return_val;
+  }
+}
+
+#pragma acc routine seq
+template <class T, class MemoryOrder, class MemoryScope>
+T device_atomic_compare_exchange(
+    T* dest, T compare, T value, MemoryOrder, MemoryScope scope) {
+  if constexpr (std::is_integral_v<T> && ((sizeof(T) == 4) || (sizeof(T) == 8))) {
```

Comment:
why only integral here, and then still cast to some other integral below?

Reply:
NVHPC supports `atomicCAS()` only for specific types of integer and floating types and does not allow casting between integer types and floating types.

Reply:
I just checked and atomicCAS does not like `int64_t` so we have to do that casting dance once way or another

Reply:
I added a comment to clarify why the code is that way

Reply:
Actually this seems to work
```diff
--- a/tpls/desul/include/desul/atomics/Compare_Exchange_OpenACC.hpp
+++ b/tpls/desul/include/desul/atomics/Compare_Exchange_OpenACC.hpp
@@ -23,7 +23,8 @@ namespace Impl {
 #pragma acc routine seq
 template <class T, class MemoryOrder, class MemoryScope>
 T device_atomic_exchange(T* dest, T value, MemoryOrder, MemoryScope /*scope*/) {
-  if constexpr (std::is_arithmetic_v<T> && ((sizeof(T) == 4) || (sizeof(T) == 8))) {
+
+  if constexpr (std::is_floating_point_v<T>) {
     T return_val;
 #pragma acc atomic capture
     {
@@ -31,6 +32,18 @@ T device_atomic_exchange(T* dest, T value, MemoryOrder, MemoryScope /*scope*/) {
       *dest = value;
     }
     return return_val;
+  } else if constexpr (std::is_arithmetic_v<T> && ((sizeof(T) == 4) || (sizeof(T) == 8))) {
+    static_assert(sizeof(unsigned int) == 4);
+    static_assert(sizeof(unsigned long long int) == 8);
+    using cas_t =
+        std::conditional_t<(sizeof(T) == 4), unsigned int, unsigned long long int>;
+    cas_t return_val;
+#pragma acc atomic capture
+    {
+      return_val = *reinterpret_cast<cas_t*>(dest);
+      *dest = value;
+    }
+    return reinterpret_cast<T&>(return_val);
   } else {
     // FIXME_OPENACC
     printf(
@@ -52,7 +65,13 @@ T device_atomic_exchange(T* dest, T value, MemoryOrder, MemoryScope /*scope*/) {
 template <class T, class MemoryOrder, class MemoryScope>
 T device_atomic_compare_exchange(
     T* dest, T compare, T value, MemoryOrder, MemoryScope scope) {
-  if constexpr (std::is_integral_v<T> && ((sizeof(T) == 4) || (sizeof(T) == 8))) {
+#ifdef DESUL_CUDA_ARCH_IS_PRE_PASCAL
+  if constexpr (std::is_same_v<T, float>) {
+#else
+  if constexpr (std::is_same_v<T, float> || std::is_same_v<T, double>) {
+#endif
+    return atomicCAS(dest, compare, value);
+  } else if constexpr ((sizeof(T) == 4) || (sizeof(T) == 8)) {
     static_assert(sizeof(unsigned int) == 4);
     static_assert(sizeof(unsigned long long int) == 8);
     using cas_t =
@@ -61,12 +80,6 @@ T device_atomic_compare_exchange(
                                  reinterpret_cast<cas_t&>(compare),
                                  reinterpret_cast<cas_t&>(value));
     return reinterpret_cast<T&>(return_val);
-#ifdef DESUL_CUDA_ARCH_IS_PRE_PASCAL
-  } else if constexpr (std::is_same_v<T, float>) {
-#else
-  } else if constexpr (std::is_same_v<T, float> || std::is_same_v<T, double>) {
-#endif
-    return atomicCAS(dest, compare, value);
   } else {
     // FIXME_OPENACC
     printf(

Reply:
But not if I drop the `is_arithmetic_v<T>` constraint in the `atomic_exchange`.
Getting an `Error: Detected unexpected atomic store opcode.`

Reply:
Will the new version work for `long double` type? 
For example, if `sizeof(long double) == 8`, then `device_atomic_compare_exchange()` will fail due to casting between `unsigned long long int` and `long double`.